### PR TITLE
Add C implementation; separate and package Python and C implementations

### PR DESCRIPTION
**Pull request by AltGr**
~~**TODO**: add tests for the C implementation! Currently there are none.~~

The C implementation follows closely the OCaml one ; all functions are in reference-passing and reference-returning style, with the return value used to denote errors on functions that may fail.

OCaml, Python and C implementations have been split in `lib_ocaml/`, `lib_python/`, `lib_c/` subdirectories.

The dune build specifications are extended to handle installation of the C artifacts and of the Python package content below `<prefix>/lib/dates_calc/`.

### CODE REVIEW

**Comment by AltGr** [comment]
@rmonat I'd be glad if you could have a quick look to check the implementation vs. the formal semantics at some point, but I'm in no hurry.

File: lib_c/dates_calc.c [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+  dc_date_round_up,
+  dc_date_round_down,
+  dc_date_round_abort
+} dc_date_rounding;
+
+
+typedef struct dc_date {
+  long int year;
+  unsigned long int month;
+  unsigned long int day;
```
Comment (rmonat) [inline lib_c/dates_calc.c:50]: Nitpicking: I think both months and days can be stored as `unsigned char`?

File: lib_c/dates_calc.c [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+void dc_copy_date(dc_date *ret, const dc_date *d) {
+  if (ret != d) {
+    ret->year = d->year;
+    ret->month = d->month;
+    ret->day = d->day;
+  }
+}
+
+/* Precondition: [1 <= d->month <= 12]. The returned day is always [1] */
+void dc_add_months(dc_date *ret, const dc_date *d, const long int months) {
```
Comment (rmonat) [inline lib_c/dates_calc.c:139]: Maybe assert the precondition?

Comment (rmonat) [inline lib_c/dates_calc.c:139]: In OCaml `add_months_to_first_of_month_date` is recursive, I think it is missing here? (ie when `|months| >= 12`)

File: lib_c/dates_calc.c [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+  if (dc_is_valid_date(d))
+    dc_copy_date(ret, d);
+  else {
+    dc_add_months (ret, d, 1);
+  }
+}
+
+dc_success dc_round_date (dc_date *ret, const dc_date_rounding rnd, const dc_date *d) {
+  if (dc_is_valid_date(d)) {
+    *ret = *d;
```
Comment (rmonat) [inline lib_c/dates_calc.c]: It seems you used copy in the other cases, why not here?

**Review by rmonat — COMMENTED** [review]
Thank you @AltGr! It seems really nice, I have a few minor comments.

**Comment by rmonat** [comment]
> **TODO**: add tests for the C implementation! Currently there are none.

On that note: I think you could do something similar to [my Python implementation](https://github.com/CatalaLang/dates-calc/blob/main/test/unit.py). The testcases are just defined in CSV files to ensure portability :)

**Comment by rmonat** [comment]
I'm not sure it's worth the extra hassle, but we may want to just have a C implementation, and then OCaml/Python modules referring to this one...

File: lib_c/dates_calc.c (hunk 1 above)
Comment (AltGr) [reply lib_c/dates_calc.c:50]: Hm, indeed, but I'll need to double-check that no intermediate calculations can overflow. Also, it might be less efficient on modern architectures.

Comment (AltGr) [reply lib_c/dates_calc.c:50]: From a reread of the code, it should be ok (we always use a larger type for the results of arithmetic on them).
I still feel safer keeping the same type rather than multiplying different integer bitlengths, though...

File: lib_c/dates_calc.c (hunk 2 above)
Comment (AltGr) [reply lib_c/dates_calc.c:139]: hm, indeed, I changed how this function works a bit and that strays away from what's been proved.
This implementation uses a direct arithmetic computation that *should* be equivalent to the recursive definition.
And it was actually wrong because it used modulo computation on months in range [1, 12] ; hopefully just shifting by -1 during the computation should ensure a valid date is returned.

File: lib_c/dates_calc.c (hunk 3 above)
Comment (AltGr) [reply lib_c/dates_calc.c]: Indeed, no good reason not to.

File: lib_c/dates_calc.c (hunk 1 above)
Comment (rmonat) [reply lib_c/dates_calc.c:50]: No worries!

**Comment by AltGr** [comment]
I took your remarks into account and implemented the CSV tests, with a couple fixes along the way (an off-by-one condition leading to a stack overflow, and a mistake on the `sscanf` interfaces). This should now be reasonably reliable.

I am not sure about this specific case, but regarding OCaml it's often convenient to have a pure implementation that doesn't require linking C code. Also, here I am afraid that, if we go the classical way of binding C functions in OCaml at least, writing the binding will be more work than the OCaml implementation (with the added risk of segfaults).

**Comment by rmonat** [comment]
Great, thanks a lot @AltGr! Feel free to do a new opam release if you require it for Catala
